io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -1103,23 +1103,25 @@ impl SnapshotOps for Lvol {
     async fn create_snapshot(
         &self,
         snap_param: SnapshotParams,
-    ) -> Result<Lvol, Error> {
+    ) -> Result<Option<Lvol>, Error> {
```

Comment:
Why not return Error::Something instead of having an option in the return? makes it rather confusing.

Reply:
> Why not return Error::Something instead of having an option in the return? makes it rather confusing.

SPDK Callback function return spdk_lvol pointer which needs to be taken forward. In the failure scenario, this pointer will be NULL. To handle both Null and valid Lvol scenario, opted for Option<Lvol> to handle the return value of the callback.
s.send((errno, None)).err(); => During Error Flow
s.send((errno, Some(lvol))).ok(); => During Success flow.

If need to handle without Option, then need to use
s.send(error, lvol) (without ok or error) and receiver will handle the error flow. but here lvol itself can be null in error flow.

Possibly, I am missing some other better approach here.

Reply:
> Why not return Error::Something instead of having an option in the return? makes it rather confusing.

Simplified the changes to handle the Null Callback without Option.

Reply:
Got your idea. 
Currently the SPDK callback was returning Null Pointer in the error flow. `Lvol::from_inner_ptr(lvol_ptr)` This was panicing if handled in the callback. To simplify it, currently the Null pointer handling is kept in the error flow and Success flow handles the valid lvol_ptr.